Add marshmallow helpers

Marenz opened the pull request.

daniel-zullo-frequenz commented on src/frequenz/quantities/_marshmallow.py:
```diff
@@ -0,0 +1,197 @@
+        unknown = EXCLUDE
+        """Ignore unknown fields."""
+
+    TYPE_MAPPING: dict[type, type[fields.Field]] = {
```
This might overwrite the `TYPE_MAPPING` from the base `Schema` class that `QuantitySchema` inherits from. You might want to extend it instead.
Alternatively you could annotate each type with `typing.Annotated`

Marenz replied:
How would the `typing.Annotated` approach work? :)

daniel-zullo-frequenz replied:
It works as a type alias with additional metadata that marshmallow can use for validation.
This is an example (I haven't tested it), I hope you'll get the idea
```python
from typing import Annotated
PercentageType = Annotated[Percentage, PercentageField]

# Then the user sets PercentageType in its Config

@dataclass
    class Config:
	my_percent_field: PercentageType = field(
        default_factory=lambda: Percentage.from_percent(25.0),
        metadata={
            "metadata": {
                "description": "Percentage to the config",
            },
        },
    ....
```

llucax replied:
> This might overwrite the `TYPE_MAPPING` from the base `Schema` class that `QuantitySchema` inherits from. You might want to extend it instead.

I don't think that is the case, at least not judging from [the example](https://github.com/lovasoa/marshmallow_dataclass?tab=readme-ov-file#custom-mapping-between-classes-and-fields). We could do some test to make sure.

The annotated approach I think is best as a quick hack, but this solution is more robust and less hacky for developers, as they don't need to declare fields with a different type than the real python type they want to use.

daniel-zullo-frequenz replied:
> I don't think that is the case, at least not judging from [the example](https://github.com/lovasoa/marshmallow_dataclass?tab=readme-ov-file#custom-mapping-between-classes-and-fields). We could do some test to make sure.

You're right. It is the right way to do it. This is just a local type mapping and it doesn't overwrite the base schema TYPE_MAPPING. Sorry for the noise!

daniel-zullo-frequenz commented on src/frequenz/quantities/_marshmallow.py:
```diff
@@ -0,0 +1,197 @@
+# License: All rights reserved
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
```
```suggestion
# Copyright © 2024 Frequenz Energy-as-a-Service GmbH
```

daniel-zullo-frequenz commented on src/frequenz/quantities/_marshmallow.py:
```diff
@@ -0,0 +1,197 @@
+            default_factory=lambda: Percentage.from_percent(25.0),
+            metadata={
+                "metadata": {
+                    "description": "Power to the config",
```
maybe Percentage instead of Power?

daniel-zullo-frequenz commented on src/frequenz/quantities/_marshmallow.py:
```diff
@@ -0,0 +1,197 @@
+            default_factory=lambda: Percentage.from_percent(25.0),
+            metadata={
+                "metadata": {
+                    "description": "Power to the config",
```
Percentage?

daniel-zullo-frequenz commented on src/frequenz/quantities/_marshmallow.py:
```diff
@@ -0,0 +1,197 @@
+        upper: the upper bound
+
+    Returns:
+        A function that will return True if the percentage is within the bounds
```
```suggestion
        Whether the percentage is within the bounds
```

Marenz replied:
No, your correction is false, it literally returns a function

shsms replied:
It is an incorrect correction, aka an incorrection.

daniel-zullo-frequenz commented on src/frequenz/quantities/_marshmallow.py:
```diff
@@ -0,0 +1,197 @@
+    }
+
+
+def within_bounds(lower: float, upper: float) -> Callable[[Percentage], bool]:
```
This function name seems to be too generic given that is specific for only `Percentage`

Marenz replied:
True, that was a copy paste from fcr

llucax commented on pyproject.toml:
```diff
@@ -81,9 +81,16 @@ dev-pytest = [
+  "frequenz-quantities[marshmallow]",
 ]
+
+marshmallow = [
+  "marshmallow == 3.22.0",
+  "marshmallow-dataclass == 8.7.1",
+]
```
These dependencies need to be wider, as this is a library, otherwise it forces to pin the dependencies to all project using this library.

Marenz replied:
Updated

llucax commented on src/frequenz/quantities/_marshmallow.py:
```diff
@@ -0,0 +1,197 @@
+        Returns:
+            The energy
+        """
+        return Energy.from_watt_hours(value)
```
I wouldn't hard code the used unit. I would actually parse from a str, so the user can set this to `100kWh` or `-10mWh`. As for serialization, just also `str()` should do the trick.

Marenz replied:
We could change that, but wouldn't wanna do that in this PR.
Besides it would require that we use strings in the config and I wasn't willing to have
`power = "10 kw"` instead of `power_kw = 10` if I can avoid it

llucax replied:
OK, I see that point, but I'm not sure. We can't enforce people to add the unit to variable names but we can enforce it in the value, so it should be less error prone, at the expense of  not being able to detect other formatting more easily with a linter (for example `power = "1;0 kw"` won't be shown as an error but `power = 1;0` probably will).

Another advantage of having the unit in the value is for cases where the range is not so clear, like timeouts, where you can end up with stuff like `time_s = 0.01` or `time_ms  = 5000` very easily, which is harder to read than `time = "10ms"` or `time = "5s"`.

But yeah, there are many trade-offs and maybe there is no clear winner, so if we add it, we could add it as an option (parse this field as a string, or as watt or as megawatt), but I agree this is for a different PR.

Marenz replied:
>we could add it as an option (parse this field as a string, or as watt or as megawatt), but I agree this is for a different PR.

well, too late,

llucax commented on src/frequenz/quantities/_marshmallow.py:
```diff
@@ -0,0 +1,197 @@
+        Returns:
+            The percentage
+        """
+        return Percentage.from_percent(value)
```
Using units here is trickier, because a ratio has no unit, and then if there is no unit is not clear if it is percentage or ratio. That said, I would also parse it as a string, and require the `%` symbol to avoid any ambiguities/confusion (so `-30%` would be valid but `-30` not).

llucax reviewed (commented):
Awesome!

You still need to expose the new classes, right? I would just make the `marshmallow` module public and that's it, so we don't mix the extension with the actual quantities code.

Marenz commented:
> You still need to expose the new classes, right?

I mean, not really, it is enough to use the `QuantitySchema`, the rest works through mapping in the schema

llucax commented:
But you still need to expose at least `QuantitySchema`! :laughing:

llucax commented:
BTW, I would expose the fields too, just in case other people want to make some advanced use and do their own mapping or something.

llucax commented on src/frequenz/quantities/_marshmallow.py:
```diff
@@ -0,0 +1,197 @@
+        A function that will return True if the percentage is within the bounds
+    """
+
+    def _within_bounds(percentage: Percentage) -> bool:
+        return lower <= percentage.as_percent() <= upper
+
+    return _within_bounds
```
BTW, is this really necessary? Can't we just use the [`Range`](https://marshmallow.readthedocs.io/en/stable/marshmallow.validate.html#marshmallow.validate.Range) validator?

```py
    @dataclass
    class Config:
        my_percent_field: Percentage = field(
            default_factory=lambda: Percentage.from_percent(25.0),
            metadata={
                "metadata": {
                    "description": "Power to the config",
                },
                "validate": Range(Percentage.zero(), Percentage.from_percentage(100)),
            },
        )
    ```

llucax commented on src/frequenz/quantities/_marshmallow.py:
```diff
@@ -0,0 +1,197 @@
+    }
+
+
+def within_bounds(lower: float, upper: float) -> Callable[[Percentage], bool]:
```
BTW, is this really necessary? Can't we just use the [`Range`](https://marshmallow.readthedocs.io/en/stable/marshmallow.validate.html#marshmallow.validate.Range) validator?

```py
    @dataclass
    class Config:
        my_percent_field: Percentage = field(
            default_factory=lambda: Percentage.from_percent(25.0),
            metadata={
                "metadata": {
                    "description": "Power to the config",
                },
                "validate": Range(Percentage.zero(), Percentage.from_percentage(100)),
            },
        )
    ```

Marenz replied:
I just tested this and yes, that seems to be an adequate replacement.

E           marshmallow.exceptions.ValidationError: {'operating_point': {'max_operating_point_pct': ['Must be greater than or equal to 0 % and less than or equal to 100 %.']}}

llucax reviewed (commented):
It would also be nice to add some tests for this.

Marenz commented:
> But you still need to expose at least QuantitySchema! 

I have! I just forgot to add that to the commit!

Marenz commented:
Updated
* Exposed each quantity field class for the power users (QUANTITY_FIELD_CLASSES)
* Added support for string based de/serialization 
* Added support for default/field based string/float serialization parameters
* Added tests
* Added support for all quantities
* Renamed to "marshmallow" for direct import
* Updated docs

llucax commented:
> Added support for string based de/serialization

Haha, so much for a different PR...

llucax commented on src/frequenz/quantities/marshmallow.py:
```diff
@@ -0,0 +1,192 @@
+        raise ValidationError("Invalid input type for QuantityField.")
+
+
+_QUANTITY_SUBCLASSES = [
+    Current,
+    Energy,
+    Frequency,
+    Percentage,
+    Power,
+    Temperature,
+    Voltage,
+]
+
+
+def _create_quantity_field_class(
+    quantity_subclass: Type[Quantity],
+) -> Type[fields.Field]:
+    """Dynamically create a QuantityField subclass for a given Quantity subclass."""
+    class_name = f"{quantity_subclass.__name__}Field"
+
+    field_class: Type[fields.Field] = type(
+        class_name,
+        (_QuantityField,),
+        {
+            "__init__": lambda self, **kwargs: super(field_class, self).__init__(
+                field_type=quantity_subclass, **kwargs
+            ),
+            "__module__": __name__,
+        },
+    )
+
+    return field_class
```
Oh, I would much rather have a LLM generating this once than generating type dynamically. For example the documentation won't be properly generated:

![image](https://github.com/user-attachments/assets/4fa86261-3ff5-4905-8130-e1bdd5d0cec9)

Marenz replied:
I get the sentiment, but.. I am pretty sure no-one really needs to use those classes directly

Marenz replied:
Like, if we want documentation for this, i would just refer people to the QuantityField class which in that case we should make public then. 

but again, I don't see any use-case why someone wants to access them.

llucax replied:
That's a fair point, but it really makes it virtually impossible to discover these classes, and makes reading the docs of the module (which is important because of `QuantitySchema`) very awkward.

To be honest if we are going to keep this code as it is, then I wouldn't export those classes at all, I think exporting them like this is basically useless and makes reading the docs about the important stuff more difficult.

Marenz replied:
I agree exporting those classes is useless :P

Marenz replied:
Though I guess that exported variable here can  be used for a users own Schema.TYPE_MAPPING, so there is that.

llucax commented on src/frequenz/quantities/marshmallow.py:
```diff
@@ -0,0 +1,192 @@
+class _QuantityField(fields.Field):
+    """Custom field for Quantity objects supporting per-field serialization configuration."""
+
+    def __init__(self, field_type: Type[Quantity], **kwargs: Any) -> None:
```
I would specify the options explicitly here (like `serialize_as_string_default`), and only leave the `**kwargs` for the options that are forwarded to the `super()` if possible. Otherwise is hard to know which options it really takes in the docs (for the parent, you can always go check the parent class docs).

Marenz replied:
That seems.. complicated?
For one, that field class and init c'tor is pretty much never used directly.

Why not keep the configuration as it is done here, through the "metadata" key
where we also configure all the other field specific things like validation, key name etc.. 

Like, I see zero reason to do this one thing differently than all the other things we can configure per-field?

llucax replied:
I guess I just don't understand this code very well. I thought this was a base class, but now I see it takes the `field_type` as argument instead. I don't know, compared to the previous version where there was a class for each field, for me this is **extremely** hard to read and understand, that's my main complain here :worried:

Marenz replied:
It took me a while as well :sweat_smile:  

This class works for all quantities.
The only thing derived classes need to do is pass their specific quantity type.

Marenz replied:
Deserialization automatically detects the type (float/str), serialization uses the schema default which can be overridden by a field-specific key.

llucax replied:
I still don't understand why, specially now that boilerplate code like this can be automatically generated with a few tabs, we want to keep dynamic unreadable code instead of having some much more obvious as self-explanatory. I see zero advantages of this approach compared to the verbose previous version and a lot of drawbacks.

Marenz replied:
> I still don't understand why

Why what? Why dynamic creation? Because that would be a different comment thread..

I am not strongly against it, I just see zero benefit as noone uses the resulting code directly..
And I doubt it makes things much clearer really.. just lots of noise

llucax replied:
From the end user PoV I'm fine with making all the fields "private" and not export them at all if we think nobody will use them. For me one or the other, but the current approach it's just exporting something that's unusable for anyone reading the docs and only adding confusion.

In terms of code, I really don't understand this code and would really prefer not to have it in the repository, again, specially when there is a super clear and easy alternative. There is no need for any black magic here, it should be very simple. The only reason I can see for it is to save a few keystrokes.

That said, I don't feel strong enough to keep investing time on it, I'm fine to merge as is as long as we don't export the fields for now, then we can see in a future PR to improve it to be able to export the fields if there is a need for it.

llucax replied:
I see I'm always late with my comments that end up in "I'm fine with merging as is" :stuck_out_tongue_closed_eyes:

llucax commented on src/frequenz/quantities/marshmallow.py:
````diff
@@ -0,0 +1,192 @@
+    ```
+    """
+
+    class Meta:
+        """Meta information."""
+
+        unknown = EXCLUDE
+        """Ignore unknown fields."""
````
Why this? I wouldn't force this option on all subclasses.

Marenz replied:
Clearly I would :P
But it's just the default, you can always pass it as parameter.
I can also remove it

llucax replied:
Yeah, this option is totally unrelated to the quantities, I would not change the default here.

llucax reviewed (commented):
Some more comments, but really cool stuff.

Marenz commented:
Removed EXCLUDE and updated documentation extensively

daniel-zullo-frequenz commented on tests/test_marshmallow.py:
```diff
@@ -0,0 +1,188 @@
+        },
+    )
+
+    @classmethod
+    def load(cls, config: dict[str, Any]) -> "Config":
+        """Load the configuration."""
+        schema = class_schema(cls, base_schema=QuantitySchema)()
+        return cast(Config, schema.load(config))
```
I think the class method can return `Self` (Self imported from typing)
```suggestion
    @classmethod
    def load(cls, config: dict[str, Any]) -> Self:
        """Load the configuration."""
        schema = class_schema(cls, base_schema=QuantitySchema)()
        return cast(Self, schema.load(config))
```

Marenz commented:
Updated: Now explicitly writing out all QuantityField subclasses
